Add Dirty Diff Peek View

pisv opened the pull request:
#### What it does

Closes #4544.

Allows the user to browse through dirty diff changes similarly to VS Code:

- `Go to Next Change`/`Go to Previous Change` (`Alt+F5`/`Shift+Alt+F5`) commands move the cursor to the next/previous change.

- `Show Next Change`/`Show Previous Change` (`Alt+F3`/`Shift+Alt+F3`) commands show a peek view with the inline diff of the next/previous change.

The peek view can also be opened by clicking on a dirty diff decoration in the gutter. Clicking on the same decoration again closes the peek view.

Basic actions for browsing through the changes and closing the peek view are provided by the peek view itself in its action bar. Other actions such as `Revert Change` and `Stage Change` can be contributed to the action bar by Theia extensions and VS Code plugins.

The implementation fully supports and has been tested with `@theia/git`:

<img width="2028" alt="Screenshot" src="https://github.com/eclipse-theia/theia/assets/8298491/5a0b7fde-e890-4c3f-9046-02bf7b45c5fe">

and `vscode.git`:

<img width="2024" alt="Screenshot" src="https://github.com/eclipse-theia/theia/assets/8298491/a6da007a-87fd-42d7-b167-7f13469d3848">

#### How to test

Verify that the following steps work correctly regardless of whether `@theia/git` or `vscode.git` is used.

0. Open a file with unstaged changes. The editor should be focused.

1. Move the cursor to the beginning of the next/previous change by using `Go to Next Change`/`Go to Previous Change` (`Alt+F5`/`Shift+Alt+F5`) commands.

2. Browse through the changes by using `Show Next Change`/`Show Previous Change` (`Alt+F3`/`Shift+Alt+F3`) commands. Press `ESC` to close the peek view.

3. Click on a dirty diff decoration in the gutter to show the corresponding change. Click on the same decoration again to close the peek view.

4. Open the peek view again and verify that each of the actions in its action bar (browsing through the changes, closing the peek view, reverting the change, staging the change) works as expected.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

tsmaeder commented:
@pisv what am I missing here? Here's what I do:

1. I changed the electron example to use the VS Code git extension because of https://github.com/eclipse-theia/theia/issues/12651
2. I open Theia on the Theia source code itself
3. I open a *.tsx file and i add "asdf" to some line
4. save

At this point, I do not get a "dirty diff" annotation in the editor gutter and I have not "Go to next Change" command availlable in command palette. Do I need to enable something?

pisv commented:
@tsmaeder Thank you for taking a look at it.

It is very strange that you don't even get dirty diff annotations in the editor gutter. Something seems to be missing indeed. It works to me locally using the following steps, based on your steps:

0. Check out the PR branch.

1. Modify the root `package.json` to remove "vscode.git" and "vscode.git-base" from "theiaPluginsExcludeIds". Modify `examples/electron/package.json` to remove "@theia/git" from "dependencies".

2. Clean-build and run Theia Electron Example:

   ```
   yarn clean
   yarn
   yarn download:plugins
   yarn electron build
   yarn electron start
   ```

3. Open the `theia` workspace.

4. Open a *.tsx file (or any other file for that matter) and modify it in any way (e.g. by adding "asdf" to some line, as you did). Saving the file should not be necessary, but it should not hurt either.

At this point, I get the proper dirty diff annotation(s) in the editor gutter and can browse through the changes using the `Alt+F5/Shift+Alt+F5` or `Alt+F3/Shift+Alt+F3` keyboard shortcuts (or with `Go -> Next Change/Previous Change` menu items).

However, I can confirm that the commands are not available in the command palette. The issue seems to be that the editor loses focus when the command palette is activated, and the commands are enabled only for the currently focused editor. I'll dig into it.

tsmaeder commented:
Here's what I do after building the merge of your branch into master (as per "command line instructions") with the git/git base extensions enabled in the electron app.

https://github.com/eclipse-theia/theia/assets/13163770/ce937f02-5140-48e0-8b8e-ba5534256ace

tsmaeder commented:
Maybe a Mac/Windows problem?

tsmaeder commented:
Doing the same workflow on a mac works. The structure of the html looks very different in both cases: "margin-view-overlays" seem to be completely absent in the Window case.

pisv commented:
> Doing the same workflow on a mac works. The structure of the html looks very different in both cases: "margin-view-overlays" seem to be completely absent in the Window case.

I see. Thank you very much for this information. It is clear now that the issue is platform-specific. I'll need some time to set up a Windows VM and debug the issue.

pisv commented:
@tsmaeder I hope that the issues found so far are fixed now.

For the sake of convenience, I pushed separate commits, which are intended to be squashed when the review is finished.

tsmaeder commented on packages/scm/src/browser/decorations/scm-decorations-service.ts:
```diff
@@ -69,7 +69,8 @@ export class ScmDecorationsService {
             try {
-                const uri = editor.uri.withScheme(currentRepo.provider.id).withQuery(`{"path":"${editor.uri['codeUri'].fsPath}","ref":"~"}`);
+                const query = { path: editor.uri['codeUri'].fsPath, ref: '~' };
+                const uri = editor.uri.withScheme(currentRepo.provider.id).withQuery(JSON.stringify(query));
```
I'll need an explanation of why this is the correct incantation. What does this do? It's obscure enough to at lease warrant a comment.

pisv replied:
It ensures that backslashes in fsPath are escaped in the query JSON, as the commit message says.

The previous version tried to construct the JSON string manually, and failed to account for the necessary escaping. Now `JSON.stringify` is used to do it properly.

pisv commented:
Converted to draft, as the dirty diff peek view no longer works when this PR is rebased on the current master. Perhaps not surprisingly, given that #13217 was merged earlier today...

tsmaeder commented:
> Perhaps not surprisingly, given that https://github.com/eclipse-theia/theia/pull/13217 was merged earlier today...

Sorry about that. Let me know if you have questions about the update.

pisv commented:
I rebased this PR on the current master, updated it to the new version of `@theia/monaco-editor-core`, and re-tested it on macOS and Windows. It is ready for review again.

tsmaeder commented:
When I hover over the little red triangle that signifies "removed line", I can get the decorations to keep flickering between the red bar and the triangle icon. Maybe there is an overlap of the decorations that gets the "hover" attribute to flip-flop between states?

pisv commented:
> When I hover over the little red triangle that signifies "removed line", I can get the decorations to keep flickering between the red bar and the triangle icon.

This issue is fixed now in the same way as it was fixed in VS Code about 6 years ago: https://github.com/microsoft/vscode/commit/00260009e8d9f3f6e7326bab3e09b1c932022e59.

Note that the issue can also be reproduced on the current master; it was not introduced by this patch.

tsmaeder reviewed (changes requested):
Thanks for this very useful PR. The main issue I have is the direct use of monaco API outside of the monaco package: this is making the task of updating to a new VS Code version more tedious. Not sure this is a direction we want to take.

pisv commented on packages/scm/src/browser/decorations/scm-decorations-service.ts:
```diff
@@ -15,64 +15,82 @@
         if (currentRepo) {
             try {
-                const uri = editor.uri.withScheme(currentRepo.provider.id).withQuery(`{"ref":"", "path":"${editor.uri.path.toString()}"}`);
+                const query = { path: editor.uri['codeUri'].fsPath, ref: '~' };
```
I'll try to explain this change in detail.

Let's start with the original line of code:

https://github.com/eclipse-theia/theia/blob/df6341a4aecec7846cf72cf9f1fdd1c2b5bc5e36/packages/scm/src/browser/decorations/scm-decorations-service.ts#L62

As far as I can tell, the format of the query in this URI is specific to `vscode.git`, and is defined by functions `fromGitUri` and `toGitUri` in [extensions/git/src/uri.ts](https://github.com/microsoft/vscode/blob/9f4bcdde684bc54e311ea49cb469771796094e7d/extensions/git/src/uri.ts). Therefore, it does not seem to support SCM providers other than `vscode.git`. I have actually installed an SVN SCM provider and can confirm that, on the current master, there are no SCM decorations shown in the editor gutter for SVN.

The patch keeps this baseline behavior, but fixes a couple of issues with the URI to ensure that it actually works in all cases with `vscode.git`:

1. As can be seen in the implementation of `toGitUri` in [extensions/git/src/uri.ts](https://github.com/microsoft/vscode/blob/9f4bcdde684bc54e311ea49cb469771796094e7d/extensions/git/src/uri.ts), `vscode.git` uses `fsPath` rather than `path` in the query.

2. As `fsPath` may contain backslashes on Windows, they need to be escaped with extra `\` in JSON. Therefore, rather than trying to construct the JSON string manually, `JSON.stringify` is now used to handle the escaping automatically, just as `vscode.git` does in its `toGitUri` function. This fixes the issue on Windows that you previously discovered.

3. `'~'` is now used as the ref instead of the empty string to ensure that it is correctly mapped to either `''` or `'HEAD'` by the `sanitizeRef` function in [extensions/git/src/fileSystemProvider.ts](https://github.com/microsoft/vscode/blob/9f4bcdde684bc54e311ea49cb469771796094e7d/extensions/git/src/fileSystemProvider.ts).

> There is code to get the file system path in file-uri.ts.

Yes, but:

1. `file-uri.ts` was node-only at the time this PR was created.

2. There is some post-processing of `fsPath` in `file-uri.ts` in the Windows case, while `vscode.git` expects the unmodified `fsPath`.

Therefore, I believe that the use of `uri['codeUri'].fsPath` is correct in this case.

tsmaeder replied:
I've looked at the corresponding code in VS Code and they seem to be using the "QuickDIffProvider" API for getting the proper URIs. There is no implementation of that concept in our Scm module as far as I can see. So this is a hack to make the feature work for git. As this is preexisting code, that's fair enough, but could you please add a comment in the code? There's just no way anyone figures this out on their own if the read this in isolation.

pisv replied:
As requested, I've added the comment in https://github.com/eclipse-theia/theia/pull/13104/commits/c98d16098d26f487399226ff727efb82ad07026b.

pisv commented:
> Thanks for this very useful PR.

Thanks for reviewing this PR.

> The main issue I have is the direct use of monaco API outside of the monaco package: this is making the task of updating to a new VS Code version more tedious. Not sure this is a direction we want to take.

Since this seems to be the main issue, let's try to address it first.

Here are my initial thoughts:

- The `debug` package already makes heavy use of the Monaco API, so there is already a precedent for the direct use of this API outside of the `monaco` package. There are other packages in Theia that also use the Monaco API directly outside of the `monaco` package, although not quite as heavily as `debug`.

- The use of the Monaco API by the `scm` package in this PR is fully encapsulated in the non-exported `DirtyDiffPeekView` class. This is in line with the first requirement set forth in the [Using code from VS Code](https://github.com/eclipse-theia/theia/blob/master/doc/vscode-usage.md) document: *Updating Monaco should not impact adopters*.

- Empirically, migration of the `DirtyDiffPeekView` class from `@theia/monaco-editor-core@1.72.3` to the new `1.83.1` version took less than a couple of hours of work in this PR, although I'm not an expert in the Monaco API in any way. In other words, it seems to satisfy the other requirement in the above mentioned document: *Adoption of a new Monaco version should generally be a straightforward, quick process*.

- I don't quite understand why the direct use of Monaco API outside of the `monaco` package would per se make the task of updating to a new version of `@theia/monaco-editor-core` more tedious.

  Would moving some code from the `scm` package down to the `monaco` package make the migration actually easier? It seems that approximately the same amount of work would still be required, as approximately the same Monaco APIs would still need to be used to implement the same functionality.

Can you explain what exactly I'm missing here?

tsmaeder commented:
@pisv I'll have time to come back to this one probably early next week.

pisv commented on packages/scm/src/browser/decorations/scm-decorations-service.ts:
```diff
@@ -15,64 +15,82 @@
+        const updateTasks = new Map<EditorWidget, () => void>();
+        this.editorManager.onCreated(editorWidget => {
+            const { editor } = editorWidget;
+            if (editor.uri.scheme !== 'file') {
```
I guess it was by analogy with:

https://github.com/eclipse-theia/theia/blob/96685e37b5450a599370355ccffd4062cf1b72ab/packages/git/src/browser/dirty-diff/dirty-diff-manager.ts#L72-L77

I agree that this restriction might be a bit too strong, although I don't have a specific case in mind when it would not hold.

However, I think that we still need to ensure that at least diff editors are ignored here, and probably some other kinds of editors too.

What would be the best way to handle it in your opinion?

tsmaeder replied:
Since the whole thing does not work for anything other than git anyway, the check is probably correct. But if we have the same concern in the `DirtyDiffManager`, why not centralize the method there?

pisv replied:
> But if we have the same concern in the `DirtyDiffManager`, why not centralize the method there?

I don't quite understand it, could you please explain what exactly I need to do here? :-)

tsmaeder replied:
In both locations, we're trying to determine whether to show the "dirty diff" for a given editor. Why not extract the check to to a method `shouldDisplayDirtyDiff(EditorWidget)`? This way way make sure the code is the same in both places and we document what the check is supposed mean.

pisv replied:
Thank you for the explanation.

I think that it would make sense to extract the check to a protected `shouldDisplayDirtyDiff(EditorWidget)` method in each of the `DirtyDiffManager` and `ScmDecorationsService` classes. It would allow clients to override those methods if necessary and would also document what the check is supposed to mean.

However, I'm not quite sure about the necessity of a shared implementation, i.e. about *make sure the code is the same in both places* part. While `DirtyDiffManager` is specific to git, `ScmDecorationsService` is clearly not, so the implementations of  the `shouldDisplayDirtyDiff(EditorWidget)` methods in those classes could potentially be quite different, although they would be the same for now. Besides, It is also unclear to me which class could contain such a shared implementation, if needed.

pisv replied:
I'd also propose `supportsDirtyDiff(EditorWidget)` instead of `shouldDisplayDirtyDiff(EditorWidget)`, if you don't mind.

pisv replied:
Addressed in https://github.com/eclipse-theia/theia/pull/13104/commits/e2857119646e3e2fc7ca11116838eccfd780f09a.

I have added `TextEditor.shouldDisplayDirtyDiff()` method. Its implementation in `MonacoEditor` returns `true`, while its implementation in `MonacoDiffEditor` returns `false`. This largely corresponds to the `isCodeEditor()` check in `DirtyDiffWorkbenchController` in [VS Code](https://github.com/microsoft/vscode/blob/d8a01ed97e89cad9a7ef5b380449ae822898c684/src/vs/workbench/contrib/scm/browser/dirtydiffDecorator.ts#L1652-L1654).

This method is called from protected `supportsDirtyDiff(TextEditor)` methods, which have been added to `DirtyDiffManager` and `ScmDecorationsService`, as proposed above.

pisv replied:
@tsmaeder Could you please take a look at this?

pisv commented on packages/scm/src/browser/scm-contribution.ts:
```diff
@@ -160,6 +193,35 @@ export class ScmContribution extends AbstractViewContribution<ScmWidget> impleme
             isEnabled: () => !!this.scmFocus.get() && !!this.acceptInputCommand()
         });
+        commandRegistry.registerCommand(SCM_COMMANDS.GOTO_NEXT_CHANGE, {
+            execute: () => this.dirtyDiffNavigator.gotoNextChange()
```
Please see https://github.com/eclipse-theia/theia/pull/13104/commits/d988fd815e23660578d5bb113efe9b8ad3e4cd92. The commit message explains the reason.

tsmaeder replied:
The problem with commit messages is that nobody sees them ;-). If we're doing something that does not make sense just from the code, we should comment in the code: otherwise the next person might "fix the problem".

pisv replied:
OK. Addressed in https://github.com/eclipse-theia/theia/pull/13104/commits/bf607eadcf272cfed4192191f2a63115f5a8a480.

pisv commented on packages/scm/src/browser/scm-tree-widget.tsx:
```diff
@@ -605,7 +605,7 @@ export class ScmResourceComponent extends ScmElement<ScmResourceComponent.Props>
     protected readonly contextMenuPath = ScmTreeWidget.RESOURCE_CONTEXT_MENU;
     protected get contextMenuArgs(): any[] {
-        if (!this.props.model.selectedNodes.some(node => ScmFileChangeNode.is(node) && node.sourceUri === this.props.sourceUri)) {
+        if (!this.props.model.selectedNodes.some(node => ScmFileChangeNode.is(node) && node === this.props.treeNode)) {
```
This is a bug fix. There can be more than one node here with the the same `sourceUri`. For example, in case of partially staged changes, there can be one node in the 'Staged Changes' group and another in the 'Changes' group, both having the same `sourceUri`.

pisv commented on packages/scm/src/browser/dirty-diff/dirty-diff-navigator.ts:
```diff
@@ -0,0 +1,288 @@
+        }
+    }
+
+    protected findNextClosestChange(line: number, inclusive: boolean): number {
```
Please look for all the references of this method to see why.

tsmaeder replied:
You're right, I missed the reference in `handleEditorMouseDown()`.

rayguncertified1 commented:
> Here's what I do after building the merge of your branch into master (as per "command line instructions") with the git/git base extensions enabled in the electron app.
> 
>  2024-02-09.11-40-14.mp4 
> In the last step, I would expect an annotation to appear where the breakpoints are in the editor.

@Luzifer

pisv commented on packages/scm/src/browser/dirty-diff/dirty-diff-widget.ts:
```diff
@@ -0,0 +1,387 @@
+            { isResizeable: true, showArrow: true, frameWidth: 1, keepEditorSelection: true, className: 'dirty-diff' },
+            StandaloneServices.get(IInstantiationService)
+        );
+        StandaloneServices.get(IThemeService).onDidColorThemeChange(this.applyTheme, this, this._disposables);
```
Addressed in https://github.com/eclipse-theia/theia/pull/13104/commits/6cdc9a38303dfa62ec54f01038f91438e49f1b97.

pisv commented on packages/scm/src/browser/dirty-diff/dirty-diff-widget.ts:
```diff
@@ -0,0 +1,387 @@
+export type DirtyDiffWidgetFactory = (props: DirtyDiffWidgetProps) => DirtyDiffWidget;
+
+@injectable()
+export class DirtyDiffWidget implements Disposable {
```
It seems I missed the connotation that "widget" has in Theia... I'm open to suggestions for a better name.

My initial proposal would be to rename the internal class `DirtyDiffPeekView` to `DirtyDiffPeekViewWidget` (since it extends `PeekViewWidget`), and then rename the exported class `DirtyDiffWidget` to `DirtyDiffPeekView`. Unfortunately, it would also affect the names of the related types and variables...

pisv commented on packages/scm/src/browser/dirty-diff/dirty-diff-widget.ts:
```diff
@@ -0,0 +1,387 @@
+}
+
+// adapted from https://github.com/microsoft/vscode/blob/823d54f86ee13eb357bc6e8e562e89d793f3c43b/extensions/git/src/staging.ts
+function applyChanges(changes: readonly ChangeRangeMapping[], original: monaco.editor.ITextModel, modified: monaco.editor.ITextModel): string {
```
> Doesn't the underling diff editor have code that would do this?

I have not found any.

> I.e. the code that is called when you press the right arrow in the middle of the split diff editor view?

When you click the right arrow, it calls `DiffEditorWidget.revert(change)`, which is not what we need here.

pisv replied:
@tsmaeder Do you think it needs any further action?

pisv commented on packages/scm/src/browser/dirty-diff/diff-computer.ts:
```diff
@@ -121,9 +158,88 @@ export interface DirtyDiff {
+    }
+}
+
+export type ChangeRangeMapping = AddedRangeMapping | RemovedRangeMapping | ModifiedRangeMapping;
```
As you wish. Addressed in https://github.com/eclipse-theia/theia/pull/13104/commits/3ef0e5a54a7b949ec635e7f567af8c5784d46271.

pisv commented on packages/scm/src/browser/dirty-diff/diff-computer.ts:
```diff
@@ -48,29 +53,49 @@ export class DiffComputer {
-                    removed.push(lastLine + 1 /* = empty line */);
+                    removed.push(currentRevisionLine + 1 /* = empty line */);
+                    if (rangeMappings) {
+                        rangeMappings.push(<RemovedRangeMapping>{ previousRange: toLineRange(change), currentRange: EmptyLineRange.afterLine(currentRevisionLine + 1) });
```
(Note that the original code also contained similar casts.)

I think in this case it acts as a safety net to guarantee that the object literal is actually what we expect here, i.e. a `RemovedRangeMapping`. It can also serve as a form of documentation...

However, I can remove these casts if you disagree.

tsmaeder replied:
It's a style issue. Up to you.

pisv commented on packages/scm/src/browser/dirty-diff/diff-computer.ts:
```diff
@@ -48,29 +53,49 @@ export class DiffComputer {
             }
         }
-        return <DirtyDiff>{ added, removed, modified };
+        return <DirtyDiff>{ added, removed, modified, rangeMappings };
```
I agree that this cast looks unnecessary. (It was in the original code.)

pisv replied:
Addressed in https://github.com/eclipse-theia/theia/pull/13104/commits/7af351fe5766cca6c43afd50ccb0602ce91381bc.

pisv commented on packages/scm/src/browser/dirty-diff/diff-computer.ts:
```diff
@@ -121,9 +158,88 @@ export interface DirtyDiff {
+
+export namespace LineRange {
+    export function isEmpty(range: LineRange | EmptyLineRange): range is EmptyLineRange {
+        return range.start === undefined || range.end === undefined;
```
Let me see. The `range` parameter is typed as `LineRange | EmptyLineRange`, so `range.start === undefined && range.end === undefined` cannot be true. Basically, it discriminates between `LineRange` and `EmptyLineRange`. I don't see why this is not correct.

pisv commented on packages/scm/src/browser/dirty-diff/diff-computer.ts:
```diff
@@ -121,9 +158,88 @@ export interface DirtyDiff {
+/**
+ * Represents a range that starts and ends either at the beginning of the {@link start} line or at the end of the {@link end} line.
+ */
+export type EmptyLineRange = { start: number; end?: undefined; } | { start?: undefined; end: number };
```
Would not that break the current semantics of the `LineRange`? Currently, a `LineRange` where `start === end` represents a single line, not an empty line range.

See also https://github.com/eclipse-theia/theia/pull/13104#discussion_r1511333972.

tsmaeder replied:
There is no promise in Theia that we won't break API. We try to be adopter-friendly in that we try to minimize breakage, but if keeping the API stable means we're making the code worse, change the API! (As a veteran of the Eclipse IDE project, I vividly recall the horrors of backward compatibility ;-)) 
In this particular case, there are a couple of factors I would consider:

1. I don't think that `LineRange' is something many adopters would rely on: it's more of an internal thing.
2. Adopting the changed semantics is straightforward
3. It probably makes the client code simpler as we don't have to deal with `undefined`: we can try this out by observing our internal client code.
4. Since the semantics were not documented before, you could argue that they're not part of the API

Assuming there's a "breaking change" notice in the `Readme.md`, I would have no problem changing this.

pisv replied:
Addressed this and [other](https://github.com/eclipse-theia/theia/pull/13104#discussion_r1495948104) concerns in https://github.com/eclipse-theia/theia/pull/13104/commits/f8c797680c437f113c38756de97f23150913daca.

I tend to agree with your arguments in this particular case, although adopting the changed semantics in `LineRange` might not be entirely "straightforward" in my experience. But let's hope that this is actually not "something many adopters would rely on".

Anyway, having lived in the Eclipse IDE land for about 20 years myself, I have come to appreciate strict rules for API evolution along with clean separation between public and internal API as one of the main strengths of Eclipse as a Platform.

I also recall a few cases where breaking even the internal API in the Eclipse Platform caused real issues downstream, including some of the Eclipse projects themselves, which forced those breaking changes to be reverted...

Do I understand it correctly that Theia currently doesn't follow such strict API policy as the Eclipse Platform?

tsmaeder replied:
Theia has two levels of extensibility: compile-time extensibility via Theia extensions and runtime-extensibility via Theia plugins (aka VS Code extensions). For runtime extensibility, we have adopted the VS Code extension API, which is stable and backward-compatible. For compile-time extensibility, we have basically no API stability guarantees. 
Since the user cannot add Theia extensions to an existing Theia-based product, this does not create problems for the end-user. Any adopter that publishes a Theia-based product will have to adapt his Theia extensions if he chooses to adopt a newer release with breaking changes. That's why we have a "breaking changes" section in the changelog and adopters are encouraged to collaborate on `docs/Migration.md` in more difficult migrations.
Personally, I find this a good compromise between ease of adoption and the dreaded ISomething1...ISomething27. Want stability => VSX API. Want total flexilibity => Theia extension. Or any mixture thereof.

pisv replied:
I see. Thank you for the detailed explanation.

tsmaeder commented:
> > Thanks for this very useful PR.
> 
> Thanks for reviewing this PR.
> 
> > The main issue I have is the direct use of monaco API outside of the monaco package: this is making the task of updating to a new VS Code version more tedious. Not sure this is a direction we want to take.
> 
> Since this seems to be the main issue, let's try to address it first.
> 
> Here are my initial thoughts:
> 
>   * The `debug` package already makes heavy use of the Monaco API, so there is already a precedent for the direct use of this API outside of the `monaco` package. There are other packages in Theia that also use the Monaco API directly outside of the `monaco` package, although not quite as heavily as `debug`.

Yes, but that is not a good thing. we have to look at every usage and determine whether it needs updating in the case of a dependency update. VS Code API is internal API that changes without notice and we have to adapt to without any guidance. 

> 
>   * The use of the Monaco API by the `scm` package in this PR is fully encapsulated in the non-exported `DirtyDiffPeekView` class. This is in line with the first requirement set forth in the [Using code from VS Code](https://github.com/eclipse-theia/theia/blob/master/doc/vscode-usage.md) document: _Updating Monaco should not impact adopters_.

That is true, but you're also not making the `PeekView' functionality available for others clients to use. For example, the comments thread widget is using a monaco editor zone widget. While the API of the zone widget implementation is arguably bad (uses VS Code API in its API), at least other Theia modules could reuse some of that implementation. 
The "Using code from VS Code" document is aimed at mitigating a bad thing. Using VS Code couples our code to an internal API we don't control and forces us to spend time on tasks which are mostly unproductive. We need to get a lot of value out of the reuse to offset the cost, and we should strive to minimize the coupling as much as we can. There are a number of places where the coupling to monaco API is not necessary. With limited resources, my aim is to clean things up if we touch an area anyway.

> 
>   * Empirically, migration of the `DirtyDiffPeekView` class from `@theia/monaco-editor-core@1.72.3` to the new `1.83.1` version took less than a couple of hours of work in this PR, although I'm not an expert in the Monaco API in any way. In other words, it seems to satisfy the other requirement in the above mentioned document: _Adoption of a new Monaco version should generally be a straightforward, quick process_.

That is true, and it's also true of the other 27 places that need to be updated in a monaco update. All these "couple of hours" add up. Updating monaco right now is a multi-week endeavor. We should strive to make it easier, not harder.

> 
>   * I don't quite understand why the direct use of Monaco API outside of the `monaco` package would per se make the task of updating to a new version of `@theia/monaco-editor-core` more tedious.
>       Would moving some code from the `scm` package down to the `monaco` package make the migration actually easier? It seems that approximately the same amount of work would still be required, as approximately the same Monaco APIs would still need to be used to implement the same functionality.

Keeping all references local to one package and consuming functionality outside that package via well-defined interfaces controllled by ourselves makes it clear what functionality we actually use from the monaco packages. I like to think in terms of "upward" and "downward" interfaces: the "upward" interface is everything VS Code offers internally. The "downward" interfaces is what Theia consumes from this "upward" interface. These interfaces are not the same and sometimes an adaptation layer is needed. Spreading this adaptation layer out makes it hard to understand and control the "downward" interface. It also makes it hard to reuse the adaptation layer, because as in the case of the `PeekView` it's usually private, because otherwise, it would make sense to make it a general service and move it to the "monaco" package.

pisv commented:
@tsmaeder I have addressed a number of smaller issues. It would be great if you could review and resolve them, or provide further feedback if necessary. This would make it easier to concentrate on the remaining ones.

tsmaeder commented:
@pisv I resolved all the comments that I don't think need further action.

pisv commented on packages/scm/src/browser/dirty-diff/diff-computer.ts:
```diff
@@ -121,9 +158,88 @@ export interface DirtyDiff {
+    /**
+     * Range mappings for the diff, if {@link DirtyDiffOptions.rangeMappings requested}.
+     */
+    readonly rangeMappings?: ChangeRangeMapping[];
```
Addressed in https://github.com/eclipse-theia/theia/pull/13104/commits/f8c797680c437f113c38756de97f23150913daca.

pisv commented on packages/scm/src/browser/dirty-diff/diff-computer.ts:
```diff
@@ -24,21 +25,25 @@ export class DiffComputer {
         const changes = this.computeDiff(previous, current);
-        let lastLine = -1;
+        let currentRevisionLine = -1;
+        let previousRevisionLine = -1;
```
I'm sorry, but I don't see how the suggested algorithm is simpler or more efficient or even much different than what is currently implemented.

For example,

```
	if (change.type === unchanged) {
	  readIndex+= change.count; 
	  writeIndex+= change.count;
	}
```

is the proposal precisely matches

```
	} else {
		// case: unchanged region
		currentRevisionLine += change.count!;
		previousRevisionLine += change.count!;
	}

```

in the current implementation, i.e. `readIndex` exactly corresponds to `previousRevisionLine`, and `writeIndex` exactly corresponds to `currentRevisionLine` in this and all the other cases.

Also, I don't think that ignoring `values` would make much if any difference. Essentially, it would be nothing more than recomputing what has already been computed.

The only significant difference that I can see is that the current implementation uses "look ahead" for coalescing, while the suggested algorithm uses "mutation" for that. That is, instead of

```
	const change = diffResult[i];
	const next = diffResult[i + 1];
	...
	} else if (change.removed && next && next.added) {
		...
		changes.push({ previousRange: toLineRange(change), currentRange: toLineRange(next) });
		currentRevisionLine += next.count!;
		previousRevisionLine += change.count!;
		i++; // consume next eagerly
```

in the current implementation, the suggested algorithm uses something like

```
	} else if (change.type === added) {
	  if (lastRangeMapping) {
		lastRangeMapping.newRange.end += change.count;
		if (lastRangeMapping?.type === removed) {
		  lastRangeMapping.type = modified;
		}
	  }
	  ...
	  writeIndex += change.count;
	}
```

I don't think that there is much difference in simplicity or efficiency between these two approaches essentially.

But I do think that the "look ahead" approach is a better fit in this particular case:

- It aligns well with the immutability of the dirty diff structures, including `LineRange`, which I believe is a good design.

- It is used not only for coalescing, but also for handling special cases such as removing at the beginning or end of the file (see references to `isNextEmptyLine`).

By the way, it is these special cases that probably make the current implementation look more complex that it actually is. And I don't think that the suggested approach would make handling of these special cases any simpler.

Finally, *if it ain't broke, don't fix it*. And I believe it ain't broke.

tsmaeder replied:
Well the point is that the special cases would go away, but I won't belabour said point.

tsmaeder commented:
Hi @pisv , except https://github.com/eclipse-theia/theia/pull/13104#discussion_r1497314715 I'm fine with this PR.

pisv commented:
(In reply to https://github.com/eclipse-theia/theia/pull/13104#issuecomment-1983041968 and https://github.com/eclipse-theia/theia/pull/13104#discussion_r1497314715)

If I understand your comments correctly, you ask me to add the necessary API to the `monaco` package in Theia and use this new API instead of using VSCode monaco API directly in `dirty-diff-widget.ts`.

After careful consideration of what this would require, I don't think I currently have enough experience in Theia to design such API properly.

I guess that was part of the reason why I decided to take the shortcut of using monaco API directly. I hoped it would not be a problem provided that such usage is well-encapsulated, as described in https://github.com/eclipse-theia/theia/pull/13104#issuecomment-1966190802.

I can understand your concern, but having invested huge amount of time in this PR already, it would probably require even more time and very detailed guidance and prompt review feedback for me to design and implement the necessary API in the desired way.

Not sure this would be an optimal direction in this case.

pisv commented:
OK. I'm trying to come up with something in the meantime. Let us see how it goes...

tsmaeder commented:
@pisv very cool, thanks. Rest assured that your efforts are much appreciated.

pisv commented:
Hi @tsmaeder,

The latest commit gets rid of monaco *internal* API use in `dirty-diff-widget.ts` by adding the necessary API to the Theia `monaco` package and using it instead. Essentially, the new API introduces two things:

* `MonacoEditorPeekViewWidget`, which is a wrapper around monaco's `PeekViewWidget`

* Support for embedded editors (i.e. editors that have a parent editor) in `MonacoEditor` and `MonacoDiffEditor`

I'm keen to know what you think about it.

tsmaeder commented:
Sounds good, I'll have a look.

tsmaeder reviewed (approved):
Looks good to me now.

pisv commented:
@tsmaeder I'm really glad to hear that you are fine with this PR now. Thank you for your review and approval.
